serve/method.go

```diff
@@ -37,45 +37,108 @@ func (m method) fullMethod() string {
 }
 
 func (m method) call(ss serverStream) error {
-	if m.desc.IsStreamingClient() || m.desc.IsStreamingServer() {
-		return status.Errorf(codes.Unimplemented, "streaming not supported: %s", m.desc.FullName())
+	if m.desc.IsStreamingClient() {
+		return m.streamingClientCall(ss)
 	}
+	return m.unaryClientCall(ss)
+}
 
+func (m method) unaryClientCall(ss serverStream) error {
+	// Handle unary client (request), with either unary or streaming server (response).
 	req := dynamicpb.NewMessage(m.desc.Input())
 	if err := ss.RecvMsg(req); err != nil {
 		return err
 	}
-	inputJSON, err := makeInputJSON(req)
+
+	input, err := makeInputJSON(req)
 	if err != nil {
 		return err
 	}
 
+	return m.evalJsonnet(input, ss)
+}
+
+func (m method) streamingClientCall(ss serverStream) error {
+	var err error
+	var input string
+	var stream []*dynamicpb.Message
+	for {
+		msg := dynamicpb.NewMessage(m.desc.Input())
+		if err = ss.RecvMsg(msg); err != nil {
+			break
+		}
+		if !m.desc.IsStreamingServer() {
+			// For a unary response, we just collect all the messages on
+			// the input stream to pass once to jsonnet.
+			stream = append(stream, msg)
+			continue
+		}
+
+		// For a streaming response, we call jsonnet once for each message
+		// on the input stream and stream out the results.
+		input, err = makeInputJSON(msg)
+		if err != nil {
+			return err
+		}
+		if err = m.evalJsonnet(input, ss); err != nil {
+			return err
+		}
+	}
+
+	if !errors.Is(err, io.EOF) {
+		return err
+	}
+
+	if !m.desc.IsStreamingServer() {
+		input, err = makeStreamingInputJSON(stream)
+		if err != nil {
+			return err
+		}
+	} else if input == "" {
+		// We are a streaming server but have nothing on the input stream.
+		// Call jsonnet once with a null request so it can choose to stream
+		// messages back if it wants.
+		input = "{response: null}"
+	}
+	return m.evalJsonnet(input, ss)
+}
+
+func (m method) evalJsonnet(input string, ss serverStream) error {
 	vm := jsonnet.MakeVM()
 	// TODO(camh): Add jsonnext.Importer
-	vm.TLACode("input", inputJSON)
-	outputJSON, err := vm.EvaluateFile(m.filename)
+	fmt.Printf("eval input  = %s\n", input)
+	vm.TLACode("input", input)
+	output, err := vm.EvaluateFile(m.filename)
 	if err != nil {
 		return err
 	}
+	fmt.Printf("eval output = %s\n", output)
 
-	resp := dynamicpb.NewMessage(m.desc.Output())
-	if err := parseOutputJSON(outputJSON, resp); err != nil {
-		return err
-	}
-	if err := ss.SendMsg(resp); err != nil {
-		return err
+	stream, err := parseOutputJSON(output, m.desc)
+	for _, resp := range stream { // stream is nil (empty) on error
```

Comment:
rather than the comment i'd check for the error, return it if not nil and return nil on the bottom.
3 lines extra, one comment less and more obvious imo.

Reply:
done